File: Session 8 - Treap For Time Series Data/Uploaded files/Java/Treap.java
import java.util.Random;

class Node {
    int ts;
    int value;
    int priority;
    Node left;
    Node right;
    Node parent;

    public Node(int ts, int value) {
        this.ts = ts;
        this.value = value;
        this.priority = (int) Math.floor(new Random().nextFloat() * 100000);
        this.left = null;
        this.right = null;
        this.parent = null;
    }
}

class BinarySearchTree {
    Node root;

    public BinarySearchTree() {
        this.root = null;
    }

    public void insert(Node node) {
        Node y = null;
        Node x = this.root;
        while (x != null) {
            y = x;
            if (node.ts < x.ts) {
                x = x.left;
            } else {
                x = x.right;
            }
        }
        node.parent = y;
        if (y == null) {
            this.root = node;
        } else if (node.ts < y.ts) {
            y.left = node;
        } else {
            y.right = node;
        }
    }

    public Node search(int key) {
        Node x = this.root;
        while (x != null && key != x.ts) {
            if (key < x.ts) {
                x = x.left;
            } else {
                x = x.right;
            }
        }
        return x;
    }

    public void inorderPrint(Node root) {
        if (root != null) {
            inorderPrint(root.left);
            System.out.print(root.ts + " ");
            inorderPrint(root.right);
        }
    }

    public void leftRotate(Node x) {
        //TODO
    }

    public void rightRotate(Node y) {
        //TODO
    }

    public int minimum(Node x) {
        while (x.left != null) {
            x = x.left;
        }
        return x.ts;
    }

    public int maximum(Node x) {
        while (x.right != null) {
            x = x.right;
        }
        return x.ts;
    }

    public int successor(Node x) {
        //TODO
        return -1;
    }

    public int predecessor(Node x) {
        //TODO
        return -1;
    }

    public void fixupInsert(Node z) {
        //TODO
    }

    public int searchOrSum(int ts1, int ts2) {
        //TODO
        return -1;
    }

    public int sumValuesBetween(Node node, int ts1, int ts2) {
        //TODO
        return -1;
    }
}

public class Main {

    public static void main(String[] args) {
        // Your code here
    }
}

File: Session 8 - Treap For Time Series Data/Uploaded files/Python/Treap.py
import math, random

class Node:
    def __init__(self, ts, value) -> None:
        self.ts = ts
        self.value = value
        self.priority = math.floor(random.random() * 100000)
        self.left = None
        self.right = None
        self.parent = None

class BinarySearchTree:
    def __init__(self) -> None:
        self.root = None
    
    def insert(self, node):
        y = None
        x = self.root
        while x is not None:
            y = x
            if node.ts < x.ts:
                x = x.left
            else:
                x = x.right
        node.parent = y
        if y is None:
            self.root = node
        elif node.ts < y.ts:
            y.left = node
        else:
            y.right = node
        
    def search(self, key):
        x = self.root
        while x is not None and key != x.ts:
            if key < x.ts:
                x = x.left
            else:
                x = x.right
        return x

    def inorder_print(self, root):
        if root is not None:
            self.inorder_print(root.left)
            print(root.ts, end=" ")
            self.inorder_print(root.right)
    
    def left_rotate(self, x):
        # TODO
        pass

    def right_rotate(self, y):
        # TODO
        pass
    
    def minimum(self, x):
        while x.left is not None:
            x = x.left
        return x.ts

    def maximum(self, x):
        while x.right is not None:
            x = x.right
        return x.ts
    
    def successor(self, x):
        # TODO
        pass

    def predecessor(self, x):
        # TODO
        pass

    def fixup_insert(self, z):
        # TODO
        pass

    def search_or_sum(self, ts1, ts2):
        # TODO
        pass

    def sum_values_between(self, node, ts1, ts2):
        # TODO
        pass
    
def main():
    # TODO
    pass    

if __name__ == '__main__':
    main()